File: pomodoro.rb
#!/usr/bin/env ruby
# pomodoro.rb
# encoding: UTF-8

require 'json'
require 'fileutils'
require 'time'

COLORS = {
  reset: "\e[0m",
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

CONFIG_FILE = File.join(Dir.home, '.pomodoro.json')
$running = false
$timer_thread = nil

def load_config
  if File.exist?(CONFIG_FILE)
    JSON.parse(File.read(CONFIG_FILE))
  else
    {
      'work_time' => 25,
      'break_time' => 5,
      'long_break_time' => 15,
      'cycles_before_long' => 4,
      'stats' => {
        'completed_pomodoros' => 0,
        'total_work_seconds' => 0,
        'total_breaks' => 0,
        'current_streak' => 0,
        'best_streak' => 0
      },
      'state' => {
        'phase' => 'idle',
        'remaining' => 0,
        'paused' => false,
        'cycle_count' => 0,
        'start_time' => nil
      }
    }
  end
end

def save_config(config)
  File.write(CONFIG_FILE, JSON.pretty_generate(config))
end

$config = load_config

def phase_name(phase)
  { 'idle' => 'Ожидание', 'work' => 'Работа', 'break' => 'Перерыв', 'long_break' => 'Длинный перерыв' }[phase] || phase
end

def phase_color(phase)
  { 'idle' => 'blue', 'work' => 'green', 'break' => 'yellow', 'long_break' => 'red' }[phase] || 'reset'
end

def format_time(seconds)
  format('%02d:%02d', seconds/60, seconds%60)
end

def play_sound
  print "\a"
end

def update_display
  state = $config['state']
  phase = phase_name(state['phase'])
  phase += ' (пауза)' if state['paused']
  print "\r#{colorize(phase, phase_color(state['phase']))} | #{colorize(format_time(state['remaining']), 'bold')}    "
end

def tick
  state = $config['state']
  if !state['paused'] && state['remaining'] > 0
    state['remaining'] -= 1
    if state['remaining'] == 0
      play_sound
      phase = state['phase']
      if phase == 'work'
        $config['stats']['completed_pomodoros'] += 1
        $config['stats']['total_work_seconds'] += $config['work_time'] * 60
        $config['stats']['current_streak'] += 1
        if $config['stats']['current_streak'] > $config['stats']['best_streak']
          $config['stats']['best_streak'] = $config['stats']['current_streak']
        end
        state['cycle_count'] += 1
        if state['cycle_count'] % $config['cycles_before_long'] == 0
          state['phase'] = 'long_break'
          state['remaining'] = $config['long_break_time'] * 60
        else
          state['phase'] = 'break'
          state['remaining'] = $config['break_time'] * 60
        end
      elsif phase == 'break' || phase == 'long_break'
        $config['stats']['total_breaks'] += 1
        state['phase'] = 'work'
        state['remaining'] = $config['work_time'] * 60
      end
      state['paused'] = false
      save_config($config)
    end
  end
  update_display
end

def timer_loop
  while $running
    tick
    sleep 1
  end
end

def main
  if ARGV.empty?
    puts "Usage: pomodoro <start|pause|resume|stop|status|set|stats|reset|quit> [args...]"
    exit 1
  end
  action = ARGV[0]
  state = $config['state']

  case action
  when 'start'
    if state['phase'] == 'idle'
      state['phase'] = 'work'
      state['remaining'] = $config['work_time'] * 60
      state['paused'] = false
      save_config($config)
    elsif state['paused']
      state['paused'] = false
      save_config($config)
    else
      puts 'Таймер уже работает.'
      exit 0
    end
    unless $running
      $running = true
      $timer_thread = Thread.new { timer_loop }
    end
  when 'pause'
    if state['phase'] == 'idle'
      puts 'Таймер не запущен.'
    elsif state['paused']
      puts 'Таймер уже на паузе.'
    else
      state['paused'] = true
      save_config($config)
      update_display
      puts "\nТаймер приостановлен."
    end
  when 'resume'
    if state['phase'] == 'idle'
      puts 'Таймер не запущен.'
    elsif !state['paused']
      puts 'Таймер не на паузе.'
    else
      state['paused'] = false
      save_config($config)
      update_display
      puts "\nТаймер возобновлён."
    end
  when 'stop'
    if state['phase'] == 'idle'
      puts 'Таймер не запущен.'
    else
      $running = false
      $timer_thread.kill if $timer_thread
      state['phase'] = 'idle'
      state['remaining'] = 0
      state['paused'] = false
      save_config($config)
      puts "\nТаймер остановлен."
    end
  when 'status'
    if state['phase'] == 'idle'
      puts 'Таймер не запущен.'
    else
      phase = phase_name(state['phase'])
      phase += ' (пауза)' if state['paused']
      puts "#{colorize(phase, phase_color(state['phase']))} | Осталось: #{colorize(format_time(state['remaining']), 'bold')}"
    end
  when 'set'
    if ARGV.size < 3
      puts 'Использование: set <param> <value>'
      exit 1
    end
    param, value = ARGV[1], ARGV[2].to_i
    if ['work_time', 'break_time', 'long_break_time', 'cycles_before_long'].include?(param)
      $config[param] = value
      save_config($config)
      puts "Параметр #{param} установлен в #{value}"
    else
      puts 'Неизвестный параметр.'
    end
  when 'stats'
    stats = $config['stats']
    puts colorize('📊 Статистика:', 'bold')
    puts "  Завершённых помидоров: #{stats['completed_pomodoros']}"
    puts "  Общее время работы: #{stats['total_work_seconds'] / 60} мин"
    puts "  Количество перерывов: #{stats['total_breaks']}"
    puts "  Текущая серия: #{stats['current_streak']}"
    puts "  Лучшая серия: #{stats['best_streak']}"
  when 'reset'
    $config['stats'] = {
      'completed_pomodoros' => 0,
      'total_work_seconds' => 0,
      'total_breaks' => 0,
      'current_streak' => 0,
      'best_streak' => 0
    }
    save_config($config)
    puts 'Статистика сброшена.'
  when 'quit'
    puts 'Выход.'
    $running = false
    $timer_thread.kill if $timer_thread
    exit 0
  else
    puts 'Неизвестная команда.'
  end
end

main if __FILE__ == $0

File: pomodoro.java
// pomodoro.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.util.*;
import com.google.gson.*; // install gson

public class pomodoro {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static class Stats {
        int completed_pomodoros;
        int total_work_seconds;
        int total_breaks;
        int current_streak;
        int best_streak;
    }

    private static class State {
        String phase;
        int remaining;
        boolean paused;
        int cycle_count;
        String start_time;
    }

    private static class Config {
        int work_time;
        int break_time;
        int long_break_time;
        int cycles_before_long;
        Stats stats;
        State state;
    }

    private static Config config;
    private static boolean running = false;
    private static Thread timerThread;
    private static String configFile = System.getProperty("user.home") + "/.pomodoro.json";

    private static void loadConfig() throws IOException {
        Gson gson = new Gson();
        Path path = Paths.get(configFile);
        if (Files.exists(path)) {
            String json = new String(Files.readAllBytes(path));
            config = gson.fromJson(json, Config.class);
        } else {
            config = new Config();
            config.work_time = 25;
            config.break_time = 5;
            config.long_break_time = 15;
            config.cycles_before_long = 4;
            config.stats = new Stats();
            config.state = new State();
            config.state.phase = "idle";
            config.state.remaining = 0;
            config.state.paused = false;
            config.state.cycle_count = 0;
        }
    }

    private static void saveConfig() throws IOException {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(config);
        Files.write(Paths.get(configFile), json.getBytes());
    }

    private static String getPhaseName(String phase) {
        switch (phase) {
            case "idle": return "Ожидание";
            case "work": return "Работа";
            case "break": return "Перерыв";
            case "long_break": return "Длинный перерыв";
            default: return phase;
        }
    }

    private static String getPhaseColor(String phase) {
        switch (phase) {
            case "idle": return BLUE;
            case "work": return GREEN;
            case "break": return YELLOW;
            case "long_break": return RED;
            default: return RESET;
        }
    }

    private static String formatTime(int seconds) {
        return String.format("%02d:%02d", seconds / 60, seconds % 60);
    }

    private static void playSound() {
        System.out.print("\007");
    }

    private static void updateDisplay() {
        State state = config.state;
        String phase = getPhaseName(state.phase);
        if (state.paused) phase += " (пауза)";
        System.out.printf("\r%s | %s    ", colorize(phase, getPhaseColor(state.phase)), colorize(formatTime(state.remaining), BOLD));
    }

    private static void tick() throws IOException {
        State state = config.state;
        if (!state.paused && state.remaining > 0) {
            state.remaining--;
            if (state.remaining == 0) {
                playSound();
                String phase = state.phase;
                if (phase.equals("work")) {
                    config.stats.completed_pomodoros++;
                    config.stats.total_work_seconds += config.work_time * 60;
                    config.stats.current_streak++;
                    if (config.stats.current_streak > config.stats.best_streak)
                        config.stats.best_streak = config.stats.current_streak;
                    state.cycle_count++;
                    if (state.cycle_count % config.cycles_before_long == 0) {
                        state.phase = "long_break";
                        state.remaining = config.long_break_time * 60;
                    } else {
                        state.phase = "break";
                        state.remaining = config.break_time * 60;
                    }
                } else if (phase.equals("break") || phase.equals("long_break")) {
                    config.stats.total_breaks++;
                    state.phase = "work";
                    state.remaining = config.work_time * 60;
                }
                state.paused = false;
                saveConfig();
            }
        }
        updateDisplay();
    }

    private static void timerLoop() {
        while (running) {
            try {
                tick();
                Thread.sleep(1000);
            } catch (InterruptedException | IOException e) {
                break;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        loadConfig();
        if (args.length == 0) {
            System.out.println("Usage: java pomodoro <start|pause|resume|stop|status|set|stats|reset|quit> [args...]");
            return;
        }
        String action = args[0];
        State state = config.state;

        switch (action) {
            case "start":
                if (state.phase.equals("idle")) {
                    state.phase = "work";
                    state.remaining = config.work_time * 60;
                    state.paused = false;
                    saveConfig();
                } else if (state.paused) {
                    state.paused = false;
                    saveConfig();
                } else {
                    System.out.println("Таймер уже работает.");
                    break;
                }
                if (!running) {
                    running = true;
                    timerThread = new Thread(() -> timerLoop());
                    timerThread.start();
                }
                break;
            case "pause":
                if (state.phase.equals("idle"))
                    System.out.println("Таймер не запущен.");
                else if (state.paused)
                    System.out.println("Таймер уже на паузе.");
                else {
                    state.paused = true;
                    saveConfig();
                    updateDisplay();
                    System.out.println("\nТаймер приостановлен.");
                }
                break;
            case "resume":
                if (state.phase.equals("idle"))
                    System.out.println("Таймер не запущен.");
                else if (!state.paused)
                    System.out.println("Таймер не на паузе.");
                else {
                    state.paused = false;
                    saveConfig();
                    updateDisplay();
                    System.out.println("\nТаймер возобновлён.");
                }
                break;
            case "stop":
                if (state.phase.equals("idle"))
                    System.out.println("Таймер не запущен.");
                else {
                    running = false;
                    if (timerThread != null) timerThread.interrupt();
                    state.phase = "idle";
                    state.remaining = 0;
                    state.paused = false;
                    saveConfig();
                    System.out.println("\nТаймер остановлен.");
                }
                break;
            case "status":
                if (state.phase.equals("idle"))
                    System.out.println("Таймер не запущен.");
                else {
                    String phase = getPhaseName(state.phase);
                    if (state.paused) phase += " (пауза)";
                    System.out.printf("%s | Осталось: %s\n", colorize(phase, getPhaseColor(state.phase)), colorize(formatTime(state.remaining), BOLD));
                }
                break;
            case "set":
                if (args.length < 3) {
                    System.out.println("Использование: set <param> <value>");
                    return;
                }
                String param = args[1];
                int val = Integer.parseInt(args[2]);
                switch (param) {
                    case "work_time": config.work_time = val; break;
                    case "break_time": config.break_time = val; break;
                    case "long_break_time": config.long_break_time = val; break;
                    case "cycles_before_long": config.cycles_before_long = val; break;
                    default: System.out.println("Неизвестный параметр."); return;
                }
                saveConfig();
                System.out.printf("Параметр %s установлен в %d\n", param, val);
                break;
            case "stats":
                Stats stats = config.stats;
                System.out.println(colorize("📊 Статистика:", BOLD));
                System.out.printf("  Завершённых помидоров: %d\n", stats.completed_pomodoros);
                System.out.printf("  Общее время работы: %d мин\n", stats.total_work_seconds / 60);
                System.out.printf("  Количество перерывов: %d\n", stats.total_breaks);
                System.out.printf("  Текущая серия: %d\n", stats.current_streak);
                System.out.printf("  Лучшая серия: %d\n", stats.best_streak);
                break;
            case "reset":
                config.stats = new Stats();
                saveConfig();
                System.out.println("Статистика сброшена.");
                break;
            case "quit":
                System.out.println("Выход.");
                running = false;
                if (timerThread != null) timerThread.interrupt();
                System.exit(0);
                break;
            default:
                System.out.println("Неизвестная команда.");
        }
    }
}
